vortex-ipc/src/reader.rs

```diff
@@ -135,84 +137,121 @@ impl<'a, R: Read> StreamArrayReader<'a, R> {
         &self.dtype
     }
 
-    // TODO(@jdcasale) return iterator over owned array instead of array
-    pub fn take(mut self, indices: &Array<'_>) -> VortexResult<OwnedArray> {
-        if !indices
-            .statistics()
-            .compute_as::<bool>(Stat::IsSorted)
-            .unwrap_or_default()
-        {
-            vortex_bail!("Indices must be sorted to take from IPC stream")
-        }
-
-        if indices
-            .statistics()
-            .compute_as_cast::<u64>(Stat::NullCount)
-            .unwrap_or_default()
-            > 0
-        {
-            vortex_bail!("Indices must not contain nulls")
-        }
+    pub fn take(self, indices: &'a Array<'_>) -> VortexResult<TakeIterator<'a, R>> {
+        if !indices.is_empty() {
+            if !indices
+                .statistics()
+                .compute_as::<bool>(Stat::IsSorted)
+                .unwrap_or_default()
+            {
+                vortex_bail!("Indices must be sorted to take from IPC stream")
+            }
 
-        if !indices.dtype().is_int() {
-            vortex_bail!("Indices must be integers")
-        }
-        if indices.dtype().is_signed_int()
-            && indices
+            if indices
                 .statistics()
-                // min cast should be safe
-                .compute_as_cast::<i64>(Stat::Min)
-                .unwrap()
-                < 0
-        {
-            vortex_bail!("Indices must be positive")
+                .compute_as_cast::<u64>(Stat::NullCount)
+                .unwrap_or_default()
+                > 0
+            {
+                vortex_bail!("Indices must not contain nulls")
+            }
+
+            if !indices.dtype().is_int() {
+                vortex_bail!("Indices must be integers")
+            }
+            if indices.dtype().is_signed_int()
+                && indices
+                    .statistics()
+                    // min cast should be safe
+                    .compute_as_cast::<i64>(Stat::Min)
```

Comment:
You can compute_min now after Rob's PR that merged this morning

Reply:
Ehh I don't think that works -- compute_min uses compute_as instead of compute_as_cast, which means that we need to know the ptype of indices (which we don't). We do know, though, that it's some unsigned int, which means it can be safely cast to i64. 

Would be nice to use compute_min, but rewriting it feels outside the scope of this pr.

Reply:
(can change the others though)